Core: Add synchronization to state changes (Fix Frame Step and FIFO Player -  Issue 8718)

EmptyChaos opened the pull request:
<p><del>[Needs PR #3769 as well]</del></p>

Alternative to #3712 that fixes the underlying race. The basic problem is that `CPU::EnableStepping` is a non-atomic function that was only designed to be called from the Host thread so it races when called from the CPU/GPU or any other system thread. One thread will go through and stop the CPU, mute the audio and pause the FIFO while the other unpauses the CPU, unmutes the audio and unpauses the FIFO resulting in a random final state (e.g. CPU running with no video, CPU and video running with no audio). Other functions which depend on `EnableStepping` like `Core::SetState` and `CPU::Break` inherit those bugs.

This PR makes the implicit design choices explicit: `CPU::EnableStepping`, `Core::SetState` and `Core::PauseAndLock` are now Host thread only. `CPU::Break` has been rewritten as a threadsafe way for system threads to do `EnableStepping(true)` only.

----

<del>I've removed the anti-deadlock hack (lock timeout) for testing purposes but that needs to go back in before this can be merged.</del> A known source of deadlocks is `wxMsgAlert` (CPU `PanicAlert`s, GUI tries to CORE_PAUSE/PauseAndLock/EnableStepping and deadlocks waiting for the CPU waiting for the GUI waiting for the CPU...), this is somewhat intractable to fix since it requires one of 1) removing wxMsgAlert entirely, 2) shifting Core::SetState/CPU::EnableStepping/Core::PauseAndLock calls onto a dedicated background thread so that the GUI never calls a blocking Core function directly, or 3) spawning an external helper process to display the messages.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3794)
<!-- Reviewable:end -->

JMC47 commented:
This looks like a great alternative to saying "HACK IT."

lioncash commented:
Review status: 0 of 28 files reviewed at latest revision, 9 unresolved discussions.

---

*[Source/Core/Core/Core.cpp, line 737 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5_uxK8ChbmK3zQwME:-KG5_uxL2diz9hfgsV1T:-804138068) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Core.cpp#L737)):*
Leftovers?

---

*[Source/Core/Core/Core.h, line 55 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5a2RRrByZBJ62T3hg:-KG5a2RRrByZBJ62T3hh:-1016615491) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Core.h#L55)):*
Spaces for comment alignment

---

*[Source/Core/Core/Movie.cpp, line 1436 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5arZrXiOwDJ9ecy_k:-KG5arZs5uS667d3y06S:1145388422) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Movie.cpp#L1436)):*
`Is a thread main` sounds kind of ambiguous with regards to meaning

---

*[Source/Core/Core/Movie.cpp, line 1458 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5b0d9CibsWpXQxnak:-KG5b0d9CibsWpXQxnal:-469765787) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Movie.cpp#L1458)):*
ditto

---

*[Source/Core/Core/HW/CPU.cpp, line 25 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5cfPE4jArqHv5bINx:-KG5cfPFP3cCOpMk1sXe:-646248409) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/HW/CPU.cpp#L25)):*
`s_cpu_step_instruction`

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 481 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5cClu1q1eIsO5hPvf:-KG5cClv6ofx8aSPtvJy:-804138068) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/HW/DVDInterface.cpp#L481)):*
Leftovers?

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 259 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5bRFq8m3iqPWqUF8U:-KG5bRFq8m3iqPWqUF8V:1317159478) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/PowerPC/PowerPC.cpp#L259)):*
spaces instead of tabs in comments for alignment

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 275 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5bSJ4XYtOxTzwnoDh:-KG5bSJ4XYtOxTzwnoDi:1317159478) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/PowerPC/PowerPC.cpp#L275)):*
spaces instead of tabs in comments for alignment

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 287 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5bMtI9rQ-txBs84_d:-KG5bMtI9rQ-txBs84_e:1317159478) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/PowerPC/PowerPC.cpp#L287)):*
spaces instead of tabs in comments for alignment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

EmptyChaos commented:
Review status: 0 of 28 files reviewed at latest revision, 9 unresolved discussions.

---

*[Source/Core/Core/Core.cpp, line 737 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5_uxK8ChbmK3zQwME:-KG5me0Djnc-U42NkPH5:-1608931054) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Core.cpp#L737)):*
It's a statement of precondition in the form of a commented assertion (Ideally it would be a real assertion instead but `Core::IsHostThread` does not exist), I'll change it to English instead of pseudo-code.

---

*[Source/Core/Core/Core.h, line 55 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5a2RRrByZBJ62T3hg:-KG5qNHn04eABgvWltwx:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Core.h#L55)):*
Done.

---

*[Source/Core/Core/Movie.cpp, line 1436 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5arZrXiOwDJ9ecy_k:-KG5sBEN6381ZPumQI8I:252946964) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Movie.cpp#L1436)):*
Changed to "entrypoint for own thread" which is less ambiguous?

---

*[Source/Core/Core/Movie.cpp, line 1458 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5b0d9CibsWpXQxnak:-KG5sI73uNFOyV7bdB5n:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/Movie.cpp#L1458)):*
Done.

---

*[Source/Core/Core/HW/CPU.cpp, line 25 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5cfPE4jArqHv5bINx:-KG5nS3odAVKcF3oeaUF:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/HW/CPU.cpp#L25)):*
Done.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 481 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5cClu1q1eIsO5hPvf:-KG5sgPY6aysye5Y0p9Y:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/HW/DVDInterface.cpp#L481)):*
Done.

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 259 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5bRFq8m3iqPWqUF8U:-KG5nwpJM_qV0VjAkUv9:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/PowerPC/PowerPC.cpp#L259)):*
Done.

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 275 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5bSJ4XYtOxTzwnoDh:-KG5nxAB_-FFRaeL2xw_:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/PowerPC/PowerPC.cpp#L275)):*
Done.

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 287 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG5bMtI9rQ-txBs84_d:-KG5o3w3dxeOdJdJL4_h:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cfd87ce0792dea1b49fdc25521ee6f0d47e306b4/Source/Core/Core/PowerPC/PowerPC.cpp#L287)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

BhaaLseN commented:
Hint, putting `squash! <start of the commit message>` or `fixup! <start of the commit message>` lets `git rebase -i` automatically move them in the right place (as long as you don't specify `--no-autosquash`, which by default is off but some clients like GitExtensions add it). Squash lets you keep both messages while Fixup discards the second one.

---

Reviewed 20 of 28 files at r1, 6 of 6 files at r2.
Review status: 26 of 28 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/Core.cpp, line 529 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6nolBksbwxKaMtGSX:-KG6nolBksbwxKaMtGSY:1459147481) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/Core.cpp#L529)):*
"That is not going to work" - why? Does it hurt including that bit in the comment?
Such half-assed comments usually always come back to bite you later, like "why is this not going to work?", and should be avoided in general (even when it may not be easy in first place)

---

*[Source/Core/Core/Core.h, line 85 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6ojcHLBjoxMJM_05h:-KG6ojcHLBjoxMJM_05i:-1094651315) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/Core.h#L85)):*
Think you could add some spaces around the `=` if you happen to come back to that file?

---

*[Source/Core/Core/HW/CPU.cpp, line 20 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6qVOe7EBphye1AHnL:-KG6qVOfg9Sc8n6LFA5B:-391362166) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/HW/CPU.cpp#L20)):*
Mh, should those have a `s_` prefix instead for being `static`?

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 36 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6qn9xndH5bz3JpKEa:-KG6qn9xndH5bz3JpKEb:1468259462) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/PowerPC/PowerPC.cpp#L36)):*
`s_` prefix?

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 307 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6r1_9-iYD6xUmo049:-KG6r1_9-iYD6xUmo04A:-1891171867) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/PowerPC/PowerPC.cpp#L307)):*
lol, but this really shouldn't be here. Have you looked into how complex deriving from `CPUCoreBase` would be?

---

*[Source/Core/DolphinWX/MainNoGUI.cpp, line 42 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6sCybs_AN0RVRhlwH:-KG6sCyc3xqWLDIfM2zQ:-694073411) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/DolphinWX/MainNoGUI.cpp#L42)):*
Have you considered using something like this inside Core (or elsewhere) rather than having the `Host_RunOnHostThread` function? One function to queue events (from any thread), another that runs on the appropriate thread and works through the list.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-:-KG6n3tsIihJbW3lgrkB:-413544566)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
@dolphin-emu-bot rebuild

EmptyChaos commented:
Review status: 7 of 27 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/Core.cpp, line 529 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6nolBksbwxKaMtGSX:-KG7tPRh-uh_JsQK9O79:-1034950659) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/Core.cpp#L529)):*
The first sentence was meant to be the explanation but I may have leaned to heavily on knowledge of the context (how the FifoPlayer works). The problem is that the FifoPlayer replaces the CPU, so what exactly is supposed to happen when `PowerPC::SingleStep` is called without a program?

The problem is no longer relevant as it's been resolved through other changes.

---

*[Source/Core/Core/Core.h, line 85 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6ojcHLBjoxMJM_05h:-KGAtoum8tRmOkk4Jbag:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/Core.h#L85)):*
Done.

---

*[Source/Core/Core/HW/CPU.cpp, line 20 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6qVOe7EBphye1AHnL:-KGAzRI2ZwrrktE-uhtY:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/HW/CPU.cpp#L20)):*
Done.

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 36 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6qn9xndH5bz3JpKEa:-KGAzYt-e9HuBLVikZhN:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/PowerPC/PowerPC.cpp#L36)):*
Done.

---

*[Source/Core/Core/PowerPC/PowerPC.cpp, line 307 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6r1_9-iYD6xUmo049:-KGAtZHmNz61e1s6x5DW:57435537) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/Core/PowerPC/PowerPC.cpp#L307)):*
I wanted to RFC the idea first. I've implemented the change now.

---

*[Source/Core/DolphinWX/MainNoGUI.cpp, line 42 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KG6sCybs_AN0RVRhlwH:-KG7tvbHgZRJYwwseTUi:-1653301165) ([raw file](https://github.com/dolphin-emu/dolphin/blob/81bf1ff12d301290aaa47fcf9d48cc126142f410/Source/Core/DolphinWX/MainNoGUI.cpp#L42)):*
Done. It's an improvement I think; it only needs a Host message (`WM_USER_DISPATCH_NEEDED`) to prompt draining of the queue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

BhaaLseN commented:
Review status: 7 of 27 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/FifoPlayer/FifoPlayer.h, line 55 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KGII2pMcOXdCbDQGoW6:-KGII2pNWbWY486TEgSE:82360200) ([raw file](https://github.com/dolphin-emu/dolphin/blob/05ff4dabc97ddbc3846881675b5364c1760e916f/Source/Core/Core/FifoPlayer/FifoPlayer.h#L55)):*
Is this `Play` comment still relevant here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

EmptyChaos commented:
Review status: 7 of 27 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/FifoPlayer/FifoPlayer.h, line 55 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KGII2pMcOXdCbDQGoW6:-KGMdf9EO3cU-k-cgqJ2:1165316422) ([raw file](https://github.com/dolphin-emu/dolphin/blob/05ff4dabc97ddbc3846881675b5364c1760e916f/Source/Core/Core/FifoPlayer/FifoPlayer.h#L55)):*
I've clarified it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
This will need a rebase

mimimi085181 commented:
Just to confirm, i can't reproduce the framestop unpause issue anymore with this. Tested with a fifo log and the intro of twilight princess(wii), where i can reproduce the issue reliably on Dolphin 4.0-9251, but not on the buiild from the buildbot based on Dolphin 4.0-9251 and this pr.

EmptyChaos commented:
I've adjusted the architecture since the original version. I elided `PowerPC::Start/Stop/Pause` into `CPU::EnableStepping/Break/Stop` and moved `PowerPC::state` into `CPU` which simplifies things since the `state` variable can now be controlled by the CPU state lock. This makes most of `CPU` threadsafe which wasn't a goal but was basically free.

Has anyone tested the Android build? I don't have a testing setup for it and the change is mildly complex.

I haven't removed `PowerPC::GetState/GetStatePtr`, they just forward to `CPU` currently. I'm undecided on what to do with it. Other than that, I believe this is ready to be reviewed now.

RisingFog commented:
This seems to need another rebase

mimimi085181 commented:
What's the status of this? This change does look really complicated. Is that what's holding it up? Should it be merged after 5.0?

@lioncash and @BhaaLseN: Do you see any issues with this?

lioncash commented:
Review status: 2 of 28 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/FifoPlayer/FifoPlayer.cpp, line 66 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW451cg-DZ_PreaYiw:-KHW451cg-DZ_PreaYix:232753735) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/FifoPlayer/FifoPlayer.cpp#L66)):*
> ```cpp
> {
> public:
> 	CPUCore(FifoPlayer* parent)
> ```

```cpp
explicit CPUCore(FifoPlayer* parent)
```

---

*[Source/Core/Core/HW/CPU.cpp, line 63 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW2FmvSLZ0iDbdfs2Q:-KHW2FmvSLZ0iDbdfs2R:1609030371) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/HW/CPU.cpp#L63)):*
> ```cpp
> 
> // Requires holding s_state_change_lock
> static inline void FlushStepSyncEventLocked()
> ```

Inline isn't necessary here, since it's static.

---

*[Source/Core/Core/HW/CPU.cpp, line 151 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW28cYVlR9EJHXSOT3:-KHW28cZS0geXxwabQ_B:-499318459) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/HW/CPU.cpp#L151)):*
> ```cpp
> 
> // Requires holding s_state_change_lock
> static inline void RunAdjacentSystems(bool running)
> ```

Ditto

---

*[Source/Core/Core/HW/CPU.cpp, line 211 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW2Lq6F_k9H0b-nEwa:-KHW2Lq6F_k9H0b-nEwb:-499318459) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/HW/CPU.cpp#L211)):*
> ```cpp
> 
> // Requires s_state_change_lock
> static inline bool SetStateLocked(State s)
> ```

Ditto

---

*[Source/Core/Core/PowerPC/PowerPC.h, line 167 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW2vIvwPUOyYSTpOCi:-KHW2vIvwPUOyYSTpOCj:2088584988) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/PowerPC/PowerPC.h#L167)):*
> ```objc
> // detect state changes.
> using CPUState = CPU::State;
> constexpr CPUState CPU_RUNNING   = CPU::CPU_RUNNING;
> ```

Is there any reason why these are here in the header instead of being called/accessed 'long-form'? (or being put in the .cpp file) PowerPC.h is used in a lot of places, and some of them don't actually need these within the namespace (especially if it's pulling things in from another namespace).

---

*[Source/Core/DolphinQt2/Main.cpp, line 26 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW4p9T9OHGKTcPAefG:-KHW4p9Ul0ivKJ8_WPK6:-806869321) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/DolphinQt2/Main.cpp#L26)):*
> ```cpp
> 	// Whenever the event loop is about to go to sleep, dispatch the jobs
> 	// queued in the Core first.
> 	QObject::connect( QAbstractEventDispatcher::instance(),
> ```

Unnecessary space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

lioncash commented:
@mimimi085181 I'm not opposed to it being merged, provided adequate testing by others for potential subtle regressions is done (it would be nice to include in 5.0, to be honest, since it'd make for a better stable release for TASers).

Some external testing from a few different people might be nice (for example, via other TASers, just to make sure it quashes all of their problems with frame stepping, etc).

<details><summary><i>Previously, mimimi085181 wrote…</i></summary>
> What's the status of this? This change does look really complicated. Is that what's holding it up? Should it be merged after 5.0?
> 
> 
> 
> @lioncash and @BhaaLseN: Do you see any issues with this?</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-:-KHW4y22gn7mXTFAhMkm:577009953)*
<!-- Sent from Reviewable.io -->

BhaaLseN commented:
I don't see much of an issue source wise, other than the ones @Lioncash already pointed out. But tbh, I only skimmed thru them and don't have as much time to spend as I'd like...

<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> @mimimi085181 I'm not opposed to it being merged, provided adequate testing by others for potential subtle regressions is done (it would be nice to include in 5.0, to be honest, since it'd make for a better stable release for TASers).
> 
> Some external testing from a few different people might be nice (for example, via other TASers, just to make sure it quashes all of their problems with frame stepping, etc).</details>


---

Review status: 2 of 28 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-:-KHWGJTgm3ulNHOmhynB:1347462175)*
<!-- Sent from Reviewable.io -->

mimimi085181 commented:
@EmptyChaos: Can you address the latest comments and rebase on top of lastest master, after pr 3810 was merged? I'd like somebody to post a build of this in the TAS community, so both TAS related changes can be tested with one build.

Dimon12321 commented:
If someone needs testing, I'll do it. I'm a TASer from TASVideos and ready to test your Work-in-Progress builds. (Sorry if my english is bad)

mimimi085181 commented:
@EmptyChaos: Thank you very much.

@Dimon12321: Now it's ready for testing. You can download the windows version of the test build from here:
https://dl.dolphin-emu.org/prs/pr-3794-dolphin-latest-x64.7z

Please test if frame advance works properly, if you set frame advance to a hotkey, and use that hotkey instead of the menu option. Also, if you have the Wii U Gamecube Controller adapter, please test with that as well.

EmptyChaos commented:
This problem is the 'gift' that keeps on giving. I found a potential race I missed earlier in `Core::PauseAndLock` so that's been partially rewritten and needs to be reviewed again.

<details><summary><i>Previously, mimimi085181 wrote…</i></summary>
> @EmptyChaos: Thank you very much.
> 
> 
> 
> @Dimon12321: Now it's ready for testing. You can download the windows version of the test build from here:
> 
> https://dl.dolphin-emu.org/prs/pr-3794-dolphin-latest-x64.7z
> 
> 
> 
> Please test if frame advance works properly, if you set frame advance to a hotkey, and use that hotkey instead of the menu option. Also, if you have the Wii U Gamecube Controller adapter, please test with that as well.</details>


---

Review status: 2 of 36 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/FifoPlayer/FifoPlayer.cpp, line 66 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW451cg-DZ_PreaYiw:-KHbL1urbBXAYshImcke:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/FifoPlayer/FifoPlayer.cpp#L66)):*
<details><summary><i>Previously, lioncash (Mathew Maidment) wrote…</i></summary>
> ```cpp
> explicit CPUCore(FifoPlayer* parent)
> ```</details>
Done.

---

*[Source/Core/Core/HW/CPU.cpp, line 63 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW2FmvSLZ0iDbdfs2Q:-KHbKX55IJ4a5ZbynZU2:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/HW/CPU.cpp#L63)):*
<details><summary><i>Previously, lioncash (Mathew Maidment) wrote…</i></summary>
> Inline isn't necessary here, since it's static.</details>
Done.

---

*[Source/Core/Core/HW/CPU.cpp, line 151 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW28cYVlR9EJHXSOT3:-KHbKXM9EJArrpmJGBa2:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/HW/CPU.cpp#L151)):*
<details><summary><i>Previously, lioncash (Mathew Maidment) wrote…</i></summary>
> Ditto</details>
Done.

---

*[Source/Core/Core/HW/CPU.cpp, line 211 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW2Lq6F_k9H0b-nEwa:-KHbKYJ5IVkTZWscrGYK:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/HW/CPU.cpp#L211)):*
<details><summary><i>Previously, lioncash (Mathew Maidment) wrote…</i></summary>
> Ditto</details>
Done.

---

*[Source/Core/Core/PowerPC/PowerPC.h, line 167 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW2vIvwPUOyYSTpOCi:-KHbKcA0H4-yPjfUtRbX:-440261452) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/Core/PowerPC/PowerPC.h#L167)):*
<details><summary><i>Previously, lioncash (Mathew Maidment) wrote…</i></summary>
> Is there any reason why these are here in the header instead of being called/accessed 'long-form'? (or being put in the .cpp file) PowerPC.h is used in a lot of places, and some of them don't actually need these within the namespace (especially if it's pulling things in from another namespace).</details>
I've removed the block and made the consumers include HW/CPU.h directly instead.

---

*[Source/Core/DolphinQt2/Main.cpp, line 26 \[r12\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KHW4p9T9OHGKTcPAefG:-KHbKus4mLnlOyTFW5Al:-995372575) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8db0f4018cbc4caa06b704d4f08b7f2704a85edf/Source/Core/DolphinQt2/Main.cpp#L26)):*
<details><summary><i>Previously, lioncash (Mathew Maidment) wrote…</i></summary>
> Unnecessary space</details>
That was for alignment. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-:-KHbL6e4Qvx2Jtj9S8gF:1421557598)*
<!-- Sent from Reviewable.io -->

Dimon12321 commented:
A very small group of people from TASVideos (and me too) have tested your build already. We don't know what else are you going to change in the build, but now Frame Advance works just fine! We got NO random unpauses. Probably RisingFog will add his word there if needed.

RisingFog commented:
What's left to be done before this is ready to merge?

EmptyChaos commented:
@Dimon12321 Thanks for testing.

@RisingFog I think (hope) this is done now, I've resolved all the races I could find. There could be more of course, but I haven't identified any other issues currently. The `MsgHandler` problem is a legacy architecture issue that will need to be fixed later.

JMC47 commented:
Everyone has been testing and reviewing this for a while and I don't know of any big flaws.  TASers have tested it, I've verified it to fix the issues I've run into, and we do know that it fixes some serious bugs and race conditions.

The question that remains is: can this affect things that we're not anticipating to cause regressions?  Considering that we're trying to gear up for 5.0 and this fixes a blocker, a decision has to be made.

If we do decide to merge it, the sooner we merge it the more likely we are to catch any of those potential unanticipated bugs.

lioncash commented:
Review status: 2 of 36 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/CPU.cpp, line 166 \[r13\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KIDnogZz7rX0OymfskD:-KIDnog_g7Sv-gAZGsNF:1315269873) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c1944f623b5918ab45152795dea9e49f57878138/Source/Core/Core/HW/CPU.cpp#L166)):*
> ```cpp
> 	s_state = CPU_POWERDOWN;
> 	s_state_cpu_cvar.notify_one();
> 	// FIXME: MsgHandler can cause this to deadlock the GUI Thread. Remove the timeout.
> ```

Now that I look over this again, it's slightly worrying to potentially include this in 5.0 with the small possibility of deadlocks (if this was pre-existing with the old code to begin with, then I don't particularly have a problem here).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

JMC47 commented:
I could deadlock the GUI thread right now through that.  I run into the bug all the time.  It's not a regression in this Pull Request.

EmptyChaos commented:
Review status: 2 of 36 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/CPU.cpp, line 166 \[r13\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KIDnogZz7rX0OymfskD:-KIFpDkG1JIP09qeDTgd:-1401184714) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c1944f623b5918ab45152795dea9e49f57878138/Source/Core/Core/HW/CPU.cpp#L166)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Now that I look over this again, it's slightly worrying to potentially include this in 5.0 with the small possibility of deadlocks (if this was pre-existing with the old code to begin with, then I don't particularly have a problem here).
</details>
Like JMC47 said, the `MsgHandler` problem is a pre-existing flaw that I don't see an easy, clean way to fix. IMHO, the best fix is to separate the GUI from the Core using an asynchronous intermediary background thread but implementing that is a large, disruptive patch in itself.

Overall, this PR handles the deadlock better than master because master lacks a timeout in `CPU::PauseAndLock` so it freezes permanently whereas this code can recover if you wait 10 seconds for the timeout.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794)*
<!-- Sent from Reviewable.io -->

lioncash commented:
Alright, given this fixes quite a few races and underlying issues, as well as making stable a much better experience for TASers (and if any regressions related to TASing happens in later dev versions, at least they have a more modern stable release to fall back to as a last resort), I'd like to include this in 5.0. Particularly because it handles a deadlock case more gracefully than what master currently does.

I'll give everyone 24 hours to leave concerns or arguments against doing this.

<details><summary><i>Previously, JMC47 wrote…</i></summary>
> I could deadlock the GUI thread right now through that.  I run into the bug all the time.  It's not a regression in this Pull Request.
</details>


---

Review status: 2 of 36 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/CPU.cpp, line 166 \[r13\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-KIDnogZz7rX0OymfskD:-KIJa5xOGP668et-HV4-:-561118624) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c1944f623b5918ab45152795dea9e49f57878138/Source/Core/Core/HW/CPU.cpp#L166)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> Like JMC47 said, the `MsgHandler` problem is a pre-existing flaw that I don't see an easy, clean way to fix. IMHO, the best fix is to separate the GUI from the Core using an asynchronous intermediary background thread but implementing that is a large, disruptive patch in itself.
> 
> Overall, this PR handles the deadlock better than master because master lacks a timeout in `CPU::PauseAndLock` so it freezes permanently whereas this code can recover if you wait 10 seconds for the timeout.
</details>
OK, that's more reassuring. I think I'd be more comfortable including this in 5.0 in that case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3794#-:-KIJaKaj2rP_blEQr7nv:-703425026)*
<!-- Sent from Reviewable.io -->

JMC47 commented:
Triggered the crash in this build; it recovered after a while seconds and I didn't lost all my progress in "Bratz: Rock Angelz"

lioncash commented:
Grace period is over with no objections.
